File: packages/hqtui/src/layout.ts
/**
 * Sizes are resolved once per frame with a single pass. No manual coordinate
 * arithmetic should ever appear in application code.
 *
 *   12       12 columns/rows
 *   "40%"    40% of the container
 *   "2fr"    two shares of whatever is left over
 *   "auto"   whatever the widget says it needs
 *   "fill"   same as "1fr"
 */
export type Size = number | string;

export interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface Constraint {
  size?: Size;
  min?: number;
  max?: number;
  /** Natural size, used by "auto" and as the floor for flexible items. */
  intrinsic?: number;
}

export const fixed = (n: number): Size => n;
export const percent = (n: number): Size => `${n}%`;
export const flex = (n = 1): Size => `${n}fr`;
export const auto: Size = "auto";
export const fill: Size = "fill";
export const remaining = (): Size => "fill";
export const minmax = (min: number, max: number): Constraint => ({ size: "fill", min, max });

interface Resolved {
  value: number;
  fr: number;
  min: number;
  max: number;
}

function parse(c: Constraint, total: number): Resolved {
  const min = c.min ?? 0;
  const max = c.max ?? Infinity;
  const size = c.size ?? "auto";
  if (typeof size === "number") return { value: size, fr: 0, min, max };
  const s = size.trim();
  if (s === "auto") return { value: c.intrinsic ?? 0, fr: 0, min, max };
  if (s === "fill") return { value: 0, fr: 1, min, max };
  if (s.endsWith("%")) {
    const pct = Number.parseFloat(s) / 100;
    return { value: Math.round(total * (Number.isFinite(pct) ? pct : 0)), fr: 0, min, max };
  }
  if (s.endsWith("fr")) {
    const n = Number.parseFloat(s);
    return { value: 0, fr: Number.isFinite(n) && n > 0 ? n : 1, min, max };
  }
  const n = Number.parseFloat(s);
  return { value: Number.isFinite(n) ? n : 0, fr: 0, min, max };
}

function clamp(v: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, v));
}

/**
 * Distribute `total` across `items`, honouring gaps, fractions, and min/max.
 * Always returns non-negative integers that sum to at most `total`.
 */
/**
 * `gap` may be one number for every seam, or one per seam. The per-seam form
 * exists so collapsed borders can be a gap of minus one between two panels and
 * the ordinary gap everywhere else on the same row.
 */
export function solve(total: number, items: Constraint[], gap: number | number[] = 0): number[] {
  const n = items.length;
  if (n === 0) return [];
  const gapTotal = Array.isArray(gap)
    ? gap.slice(0, n - 1).reduce((sum, value) => sum + value, 0)
    : gap * (n - 1);
  const available = Math.max(0, total - gapTotal);
  const parsed = items.map((c) => parse(c, available));

  let used = 0;
  let frTotal = 0;
  const out = new Array<number>(n).fill(0);

  for (let i = 0; i < n; i++) {
    const p = parsed[i];
    if (p.fr > 0) {
      frTotal += p.fr;
      out[i] = -1; // resolved in the flexible pass
    } else {
      out[i] = clamp(Math.round(p.value), p.min, Math.min(p.max, available));
      used += out[i];
    }
  }

  let free = Math.max(0, available - used);
  if (frTotal > 0) {
    // Two passes: clamped items give their surplus back to the rest.
    const flexible: number[] = [];
    for (let i = 0; i < n; i++) if (out[i] === -1) flexible.push(i);

    let remainingFr = frTotal;
    let pool = free;
    const pending = new Set(flexible);
    let changed = true;
    while (changed && pending.size > 0) {
      changed = false;
      for (const i of [...pending]) {
        const p = parsed[i];
        const share = remainingFr > 0 ? (pool * p.fr) / remainingFr : 0;
        const clamped = clamp(share, p.min, p.max);
        if (clamped !== share) {
          out[i] = Math.round(clamped);
          pool -= out[i];
          remainingFr -= p.fr;
          pending.delete(i);
          changed = true;
        }
      }
    }
    // Distribute what is left, giving rounding remainder to the largest shares.
    let assigned = 0;
    const order = [...pending];
    order.forEach((i, k) => {
      const p = parsed[i];
      const exact = remainingFr > 0 ? (pool * p.fr) / remainingFr : 0;
      const v = k === order.length - 1 ? Math.max(0, pool - assigned) : Math.floor(exact);
      out[i] = v;
      assigned += v;
    });
    free = 0;
  }

  // Overflow: shrink from the end until it fits rather than drawing outside.
  let sum = out.reduce((a, b) => a + b, 0);
  if (sum > available) {
    for (let i = n - 1; i >= 0 && sum > available; i--) {
      const shrink = Math.min(out[i] - parsed[i].min, sum - available);
      if (shrink > 0) {
        out[i] -= shrink;
        sum -= shrink;
      }
    }
    for (let i = n - 1; i >= 0 && sum > available; i--) {
      const shrink = Math.min(out[i], sum - available);
      out[i] -= shrink;
      sum -= shrink;
    }
  }

  return out;
}

/**
 * Where leftover space goes.
 *
 * It only ever applies when there is slack, and a container holding any `fr`
 * or `fill` child has none -- that child has already absorbed it. So this is
 * inert exactly where it would otherwise fight with the constraints.
 */
export type Justify =
  | "start"
  | "end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

/**
 * How much slack sits before item `i`, as an exact fraction.
 *
 * Every mode is a different answer to that one question, which is why they can
 * share the rounding below rather than each doing their own and each getting a
 * different kind of off-by-one.
 */
function before(i: number, slack: number, count: number, justify: Justify): number {
  switch (justify) {
    case "end":
      return slack;
    case "center":
      // Floor, so an odd cell falls after the content rather than before it.
      // That is what CSS and ratatui do, and it is what reads as centred.
      return Math.floor(slack / 2);
    case "space-between":
      // One child has nothing to sit between, so it stays where it started.
      return count > 1 ? (i * slack) / (count - 1) : 0;
    case "space-evenly":
      return ((i + 1) * slack) / (count + 1);
    case "space-around":
      // Half a share at each end, a whole one between.
      return ((i + 0.5) * slack) / count;
    default:
      return 0;
  }
}

/**
 * The offset before the first child, and the extra added at each seam.
 *
 * Cells are whole, and rounding each gap on its own loses a cell here and
 * gains one there. Rounding the *cumulative* offset instead and taking
 * differences means the parts always add up to exactly the slack, whatever the
 * mode and however awkward the division.
 */
export function distribute(
  slack: number,
  count: number,
  justify: Justify,
): { lead: number; seams: number[] } {
  const seams = new Array<number>(Math.max(0, count - 1)).fill(0);
  if (slack <= 0 || count === 0 || justify === "start") return { lead: 0, seams };

  const at = (i: number): number => Math.round(before(i, slack, count, justify));
  for (let i = 0; i + 1 < count; i++) seams[i] = at(i + 1) - at(i);
  return { lead: at(0), seams };
}

/** Lay children out along one axis inside `rect`. */
export function stack(
  rect: Rect,
  items: Constraint[],
  direction: "row" | "column",
  gap: number | number[] = 0,
  justify: Justify = "start",
): Rect[] {
  const horizontal = direction === "row";
  const axis = horizontal ? rect.width : rect.height;
  const sizes = solve(axis, items, gap);
  const seam = (i: number) => (Array.isArray(gap) ? (gap[i] ?? 0) : gap);

  let gapTotal = 0;
  for (let i = 0; i + 1 < sizes.length; i++) gapTotal += seam(i);
  const slack = Math.max(0, axis - sizes.reduce((a, b) => a + b, 0) - gapTotal);
  const { lead, seams } = distribute(slack, sizes.length, justify);

  const out: Rect[] = [];
  let offset = (horizontal ? rect.x : rect.y) + lead;
  sizes.forEach((size, i) => {
    out.push(
      horizontal
        ? { x: offset, y: rect.y, width: size, height: rect.height }
        : { x: rect.x, y: offset, width: rect.width, height: size },
    );
    offset += size + seam(i) + (seams[i] ?? 0);
  });
  return out;
}

export type Padding = number | [number, number] | [number, number, number, number] |
  { top?: number; right?: number; bottom?: number; left?: number };

export function normalizePadding(p: Padding | undefined): [number, number, number, number] {
  if (p === undefined) return [0, 0, 0, 0];
  if (typeof p === "number") return [p, p, p, p];
  if (Array.isArray(p)) {
    if (p.length === 2) return [p[0], p[1], p[0], p[1]];
    return [p[0], p[1], p[2], p[3]];
  }
  return [p.top ?? 0, p.right ?? 0, p.bottom ?? 0, p.left ?? 0];
}

/** Shrink a rect by padding, never past zero. */
export function inset(rect: Rect, padding: Padding): Rect {
  const [t, r, b, l] = normalizePadding(padding);
  return {
    x: rect.x + l,
    y: rect.y + t,
    width: Math.max(0, rect.width - l - r),
    height: Math.max(0, rect.height - t - b),
  };
}

export function intersect(a: Rect, b: Rect): Rect {
  const x = Math.max(a.x, b.x);
  const y = Math.max(a.y, b.y);
  const x2 = Math.min(a.x + a.width, b.x + b.width);
  const y2 = Math.min(a.y + a.height, b.y + b.height);
  return { x, y, width: Math.max(0, x2 - x), height: Math.max(0, y2 - y) };
}

export function contains(rect: Rect, x: number, y: number): boolean {
  return x >= rect.x && y >= rect.y && x < rect.x + rect.width && y < rect.y + rect.height;
}

export function isEmpty(rect: Rect): boolean {
  return rect.width <= 0 || rect.height <= 0;
}

File: ports/python/hqtui/layout.py
"""Sizes are resolved once per frame with a single pass.

No manual coordinate arithmetic should ever appear in application code.

::

    12       12 columns/rows
    "40%"    40% of the container
    "2fr"    two shares of whatever is left over
    "auto"   whatever the widget says it needs
    "fill"   same as "1fr"

Python takes the reference implementation's spelling directly: a ``Size`` is an
``int`` or one of those strings, and nothing needs constructing.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from enum import Enum
from typing import NamedTuple, Sequence

from .color import round_half_up

__all__ = [
    "Constraint",
    "Direction",
    "Padding",
    "Rect",
    "Size",
    "normalize_padding",
    "solve",
    "stack",
]

Size = "int | str"
"""A number of cells, or one of ``"auto"``, ``"fill"``, ``"40%"``, ``"2fr"``."""

Padding = "int | tuple[int, int] | tuple[int, int, int, int]"
"""One value, ``(vertical, horizontal)``, or ``(top, right, bottom, left)``."""


@dataclass(frozen=True, slots=True)
class Constraint:
    """What one item contributes to a layout solve."""

    size: "int | str | None" = None
    min: int | None = None
    max: int | None = None
    intrinsic: int | None = None
    """Natural size, used by ``"auto"`` and as the floor for flexible items."""


class Rect(NamedTuple):
    """A rectangle in absolute buffer coordinates."""

    x: int
    y: int
    width: int
    height: int

    @property
    def is_empty(self) -> bool:
        return self.width <= 0 or self.height <= 0

    def contains(self, x: int, y: int) -> bool:
        return self.x <= x < self.x + self.width and self.y <= y < self.y + self.height

    def inset(self, padding: "int | tuple") -> "Rect":
        """Shrink by padding, never past zero."""
        t, r, b, left = normalize_padding(padding)
        return Rect(
            self.x + left,
            self.y + t,
            max(0, self.width - left - r),
            max(0, self.height - t - b),
        )

    def intersect(self, other: "Rect") -> "Rect":
        x, y = max(self.x, other.x), max(self.y, other.y)
        x2 = min(self.x + self.width, other.x + other.width)
        y2 = min(self.y + self.height, other.y + other.height)
        return Rect(x, y, max(0, x2 - x), max(0, y2 - y))


def normalize_padding(padding: "int | tuple | None") -> tuple[int, int, int, int]:
    if padding is None:
        return (0, 0, 0, 0)
    if isinstance(padding, int):
        return (padding, padding, padding, padding)
    if len(padding) == 2:
        v, h = padding
        return (v, h, v, h)
    return tuple(padding)  # type: ignore[return-value]


class Direction(str, Enum):
    """Which way a container lays its children out."""

    ROW = "row"
    COLUMN = "column"


class _Resolved(NamedTuple):
    value: float
    fr: float
    min: float
    max: float


def _leading_float(s: str) -> float:
    """``Number.parseFloat``: read as much of a leading number as parses."""
    s = s.strip()
    end, seen_digit, seen_dot = 0, False, False
    while end < len(s):
        c = s[end]
        if c in "+-" and end == 0:
            pass
        elif c.isdigit():
            seen_digit = True
        elif c == "." and not seen_dot:
            seen_dot = True
        else:
            break
        end += 1
    if not seen_digit:
        return 0.0
    try:
        return float(s[:end])
    except ValueError:
        return 0.0


def _parse(c: Constraint, total: int) -> _Resolved:
    lo = 0.0 if c.min is None else float(c.min)
    hi = math.inf if c.max is None else float(c.max)
    size = "auto" if c.size is None else c.size

    if isinstance(size, (int, float)) and not isinstance(size, bool):
        return _Resolved(float(size), 0.0, lo, hi)

    s = str(size).strip()
    if s == "auto":
        return _Resolved(float(c.intrinsic or 0), 0.0, lo, hi)
    if s == "fill":
        return _Resolved(0.0, 1.0, lo, hi)
    if s.endswith("%"):
        pct = _leading_float(s[:-1]) / 100
        if not math.isfinite(pct):
            pct = 0.0
        return _Resolved(round_half_up(total * pct), 0.0, lo, hi)
    if s.endswith("fr"):
        n = _leading_float(s[:-2])
        return _Resolved(0.0, n if math.isfinite(n) and n > 0 else 1.0, lo, hi)
    return _Resolved(_leading_float(s), 0.0, lo, hi)


def _clamp(v: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, v))


def solve(
    total: int, items: Sequence[Constraint], gap: "int | Sequence[int]" = 0
) -> list[int]:
    """Distribute ``total`` across ``items``, honouring gaps, fractions and
    min/max. Always returns non-negative sizes that sum to at most ``total``.

    ``gap`` may be one number for every seam, or one per seam. The per-seam form
    exists so collapsed borders can be a gap of minus one between two panels and
    the ordinary gap everywhere else on the same row.
    """
    n = len(items)
    if n == 0:
        return []
    gaps = list(gap) if isinstance(gap, (list, tuple)) else [gap] * (n - 1)
    available = max(0, total - sum(gaps[: n - 1]))
    parsed = [_parse(c, available) for c in items]

    used = 0
    fr_total = 0.0
    # -1 marks an item resolved in the flexible pass below.
    out = [-1] * n

    for i in range(n):
        p = parsed[i]
        if p.fr > 0:
            fr_total += p.fr
        else:
            out[i] = int(_clamp(round_half_up(p.value), p.min, min(p.max, available)))
            used += out[i]

    free = max(0.0, float(available - used))
    if fr_total > 0:
        # Two passes: clamped items give their surplus back to the rest.
        remaining_fr = fr_total
        pool = free
        pending = [i for i in range(n) if out[i] == -1]

        changed = True
        while changed and pending:
            changed = False
            for i in list(pending):
                p = parsed[i]
                share = pool * p.fr / remaining_fr if remaining_fr > 0 else 0.0
                clamped = _clamp(share, p.min, p.max)
                if clamped != share:
                    out[i] = int(round_half_up(clamped))
                    pool -= out[i]
                    remaining_fr -= p.fr
                    pending.remove(i)
                    changed = True

        # Distribute what is left, giving the rounding remainder to the last.
        assigned = 0
        for k, i in enumerate(pending):
            p = parsed[i]
            exact = pool * p.fr / remaining_fr if remaining_fr > 0 else 0.0
            v = max(0, int(pool) - assigned) if k == len(pending) - 1 else int(math.floor(exact))
            out[i] = v
            assigned += v

    # Overflow: shrink from the end until it fits rather than drawing outside.
    total_out = sum(out)
    if total_out > available:
        for i in range(n - 1, -1, -1):
            if total_out <= available:
                break
            shrink = min(out[i] - int(parsed[i].min), total_out - available)
            if shrink > 0:
                out[i] -= shrink
                total_out -= shrink
        for i in range(n - 1, -1, -1):
            if total_out <= available:
                break
            shrink = min(out[i], total_out - available)
            out[i] -= shrink
            total_out -= shrink

    return [max(0, v) for v in out]


#: Where leftover space goes.
#:
#: It only ever applies when there is slack, and a container holding any ``fr``
#: or ``fill`` child has none -- that child has already absorbed it. So this is
#: inert exactly where it would otherwise fight with the constraints.
JUSTIFY = ("start", "end", "center", "space-between", "space-around", "space-evenly")


def _before(i: int, slack: float, count: int, justify: str) -> float:
    """How much slack sits before item ``i``, as an exact fraction.

    Every mode is a different answer to that one question, which is why they
    share the rounding below rather than each growing their own off-by-one.
    """
    if justify == "end":
        return slack
    if justify == "center":
        # Floor, so an odd cell falls after the content rather than before it.
        return float(int(slack // 2))
    if justify == "space-between":
        return (i * slack / (count - 1)) if count > 1 else 0.0
    if justify == "space-evenly":
        return (i + 1) * slack / (count + 1)
    if justify == "space-around":
        return (i + 0.5) * slack / count
    return 0.0


def _round_half_up(value: float) -> int:
    """Python rounds halves to even; every other port rounds them up."""
    return int(math.floor(value + 0.5))


def distribute(slack: int, count: int, justify: str = "start") -> tuple[int, list[int]]:
    """The offset before the first child, and the extra added at each seam.

    Cells are whole, and rounding each gap on its own loses one here and gains
    one there. Rounding the cumulative offset and taking differences means the
    parts always add up to exactly the slack.
    """
    seams = [0] * max(0, count - 1)
    if slack <= 0 or count == 0 or justify == "start":
        return 0, seams
    at = lambda i: _round_half_up(_before(i, float(slack), count, justify))  # noqa: E731
    for i in range(max(0, count - 1)):
        seams[i] = max(0, at(i + 1) - at(i))
    return at(0), seams


def stack(
    rect: Rect,
    items: Sequence[Constraint],
    direction: "Direction | str" = Direction.COLUMN,
    gap: "int | Sequence[int]" = 0,
    justify: str = "start",
) -> list[Rect]:
    """Lay children out along one axis inside ``rect``."""
    horizontal = direction == Direction.ROW
    axis = rect.width if horizontal else rect.height
    sizes = solve(axis, items, gap)
    gaps = list(gap) if isinstance(gap, (list, tuple)) else [gap] * max(0, len(sizes) - 1)

    gap_total = sum(gaps[i] for i in range(max(0, len(sizes) - 1)) if i < len(gaps))
    slack = max(0, axis - sum(sizes) - gap_total)
    lead, extra = distribute(slack, len(sizes), justify)

    out: list[Rect] = []
    offset = (rect.x if horizontal else rect.y) + lead
    for i, size in enumerate(sizes):
        if horizontal:
            out.append(Rect(offset, rect.y, size, rect.height))
        else:
            out.append(Rect(rect.x, offset, rect.width, size))
        seam = (gaps[i] if i < len(gaps) else 0) + (extra[i] if i < len(extra) else 0)
        offset += size + seam
    return out
